piker/brokers/kraken.py

```diff
@@ -162,6 +222,129 @@ async def _public(
         )
         return resproc(resp, log)
 
+    async def _private(
+        self,
+        method: str,
+        data: dict,
+        uri_path: str
+    ) -> Dict[str, Any]:
+        headers = {
+            'Content-Type': 
+                'application/x-www-form-urlencoded',
+            'API-Key':
+                self._api_key,
+            'API-Sign':
+                get_kraken_signature(uri_path, data, self._secret)
+        }
+        resp = await self._sesh.post(
+            path=f'/private/{method}',
+            data=data,
+            headers=headers,
+            timeout=float('inf')
+        )
+        return resproc(resp, log)
+
+    async def kraken_endpoint(
+        self,
+        method: str,
+        data: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        uri_path = f'/0/private/{method}'
+        data['nonce'] = str(int(1000*time.time()))
+        resp = await self._private(method, data, uri_path)
+        return resp
+
+    async def get_positions(
+        self,
+        data: Dict[str, Any] = {}
+    ) -> Dict[str, Any]:
+        data['ofs'] = 0
+        positions = {}
+        vols = {}
+        # Grab all trade history
+        while True:
```

Comment:
not sure if there's some kind of state you're keeping here (`ofs` ?) but in theory you can do these requests concurrently and speedup the startup by doing a batch request for all trades (starting at different counts?) and then a single logic block to the aggregation calcs/work.

Reply:
oh yeah looks like you keep passing `data` so i would figure out if you can do this batch-wise though it might be tricky to figure out *the end* trade given that you seem to be checking `resp['result']['trades'] == {}`?

Reply:
Ok let's postpone and make an issue for this for now.

we can check how `cryptofeed` does this and maybe just delegate to them?

Reply:
`cryptofeed` point's of interest:
- [trades parsing mixin method](https://github.com/bmoscon/cryptofeed/blob/master/cryptofeed/exchanges/mixins/kraken_rest.py#L112), which calls
- [`._historical_trades()](https://github.com/bmoscon/cryptofeed/blob/c37f9c4fc19cafa877eda8ee5ceb2c1c3c17a830/cryptofeed/exchanges/mixins/kraken_rest.py#L127) which has some very [interesting edge case handling](https://github.com/bmoscon/cryptofeed/blob/c37f9c4fc19cafa877eda8ee5ceb2c1c3c17a830/cryptofeed/exchanges/mixins/kraken_rest.py#L133) which seems to use some kinda `&since=` http params flag? I dunno.

Reply:
ah sorry that was the wrong endpoint.

You actually probably want this [`.trade_history()` method](https://github.com/bmoscon/cryptofeed/blob/c37f9c4fc19cafa877eda8ee5ceb2c1c3c17a830/cryptofeed/exchanges/mixins/kraken_rest.py#L230) which is private and uses the same actual web ep.

Reply:
yeah [i don't think they compute pps whatsoever](https://github.com/bmoscon/cryptofeed/search?q=position+kraken) in `cryptofeed` so we're going to need this implementation either way 🏄🏼